### PR TITLE
Add event model protocol

### PR DESCRIPTION
**Pull request by MartinHjelmare**
- Here's my take on library design model.
- Please comment and criticize!
- I've included just enough to show the concept I hope, based on the existing code.
~~- Names are hard. Eg I swapped Data for Gateway, but not sure about that one.~~

## Outline

- The model package represents the pieces that together make up the Z-Wave network, controller and sent data.
- The protocol enum and event handlers handles the actions of each model piece. What should happen when a message or event comes in to a piece in the model.

### CODE REVIEW

**Comment by marcelveldt** [comment]
Can we settle on what to do with this ? I'm about to code the foundation in the integration and it's more or less dependent of choices made here, especially the event listeners...

**Comment by MartinHjelmare** [comment]
I don't think my PR will change event listeners or event emission. It's just a matter of how we abstract the internals of the library at the moment.

**Comment by marcelveldt** [comment]
No, but we can add that after. For example an event when the initial state is received would be great.
I'd like to add that but maybe better to first merge this one so we have a solid foundation to work from.

**Comment by MartinHjelmare** [comment]
What could change things is if we introduce an additional model layer, but since we should map against zwave-js model we shouldn't do that.

**Comment by MartinHjelmare** [comment]
I've now rebased on the updated model in master branch and applied this PR suggestion to that. I haven't built out the node and value protocols yet. I'll do that if we want to go ahead with this.

File: test/model/test_driver.py
Comment (MartinHjelmare) [inline test/model/test_driver.py:18]: There seems to be a missing key in the fixture data we use here. We get a `KeyError: 'source'`.

Comment (MartinHjelmare) [reply test/model/test_driver.py:20]: Nevermind. It was just me.

Comment (marcelveldt) [reply test/model/test_driver.py:20]: > I've now rebased on the updated model in master branch and applied this PR suggestion to that. I haven't built out the node and value protocols yet. I'll do that if we want to go ahead with this.

Looks good!. I could really use this merged in for the integration as I can reference the enums etc.

File: test/model/test_driver.py
Comment (balloob) [inline test/model/test_driver.py:17]: These are not driver events. These are all events stored in the dump, which can be any source.

File: zwave_js_server/model/controller.py
Comment (balloob) [inline zwave_js_server/model/controller.py:11]: We can't distinguish the events in different classes. They all come in at at the root and are forwarded accordingly. They all follow the same format.

What's the benefit from breaking it up into different classes instead of just having an event class?

File: zwave_js_server/protocol/__init__.py
Comment (balloob) [inline zwave_js_server/protocol/__init__.py:8]:
```suggestion
    type: str
```

File: zwave_js_server/protocol/__init__.py
Comment (balloob) [inline zwave_js_server/protocol/__init__.py:19]: This could be just on a protocol instance.

File: zwave_js_server/protocol/controller.py
Comment (balloob) [inline zwave_js_server/protocol/controller.py:26]: I see what's going on here, but it is this really improving the existing logic? I think it's harder to follow. 

Instead of having a Controller instance that has controller properties and methods, we know move the methods to a second class and have that class mutate the public instance variables of a controller? That kinda treats a controller like a C-like struct: just a holder of data. 

These methods are changing instance variables of the models and `event.data`. It fels weird to decouple that. A protocol mutates data and the owner class is not in the loop?

Comment (balloob) [reply zwave_js_server/protocol/controller.py:26]: I think that we can decide to have the `get_handler` logic be included in the models that receive events so that we can still split out handling of each event into different methods.

Comment (MartinHjelmare) [reply zwave_js_server/protocol/controller.py:26]: I think that works.

File: zwave_js_server/model/controller.py
Comment (MartinHjelmare) [reply zwave_js_server/model/controller.py:11]: We can use a single class. This is a leftover from before we settled on mapping the models 1:1 to Z-Wave JS.

File: test/model/test_driver.py
Comment (MartinHjelmare) [reply test/model/test_driver.py:17]: I'll make a common Event class.

File: zwave_js_server/event.py
Comment (MartinHjelmare) [inline zwave_js_server/event.py:7]: Should we rename this to EventModelBase?

File: zwave_js_server/protocol/__init__.py
Comment (MartinHjelmare) [reply zwave_js_server/protocol/__init__.py]: I've moved this into the EventBase as a helper.

File: zwave_js_server/protocol/controller.py
Comment (MartinHjelmare) [reply zwave_js_server/protocol/controller.py]: I've moved the handlers to each model class.

**Comment by MartinHjelmare** [comment]
The feedback really cleaned things up.

**Comment by MartinHjelmare** [comment]
Converting to draft so we don't merge by mistake.

File: zwave_js_server/event.py
Comment (MartinHjelmare) [inline zwave_js_server/event.py:17]: I'm not sure how to type the protocol in the best way since we want to be able to accept any model protocol Enum.

File: zwave_js_server/event.py
Comment (balloob) [inline zwave_js_server/event.py:42]: I think that we can do it without a protocol. We can just check for the existence of the methods.

```suggestion
    handler = getattr(self, f"handle_{event.type.replace(' ', '_')}", None)
    if handler is None:
        LOGGER.debug("Received unknown event")
        return
    handler(event)
```

File: zwave_js_server/model/driver.py
Comment (balloob) [inline zwave_js_server/model/driver.py:30]: We don't forward these. The API only comes online once the driver is ready.

Comment (MartinHjelmare) [reply zwave_js_server/model/driver.py:30]: Is it just this event ("driver ready") that we don't handle? We handle "error" and "all nodes ready"?

Comment (balloob) [reply zwave_js_server/model/driver.py:30]: `all nodes ready` we do. Error we don't forward because it's unclear to me when we get it. Is it during initialization or runtime too

**Comment by marcelveldt** [comment]
you should probably merge this before it gets too large and address any issue that may still be there in follow up PR's if needed. Overall it's looking good!

**Comment by MartinHjelmare** [comment]
Just waiting for Paulus approval.

**Review by balloob — APPROVED** [review]
Nice! 🎉

**Pull request merged**
